File: sdk/src/main/java/com/libreshockwave/cast/ScriptType.java
package com.libreshockwave.cast;

/**
 * Types of Lingo scripts.
 */
public enum ScriptType {
    INVALID(0, "invalid"),
    SCORE(1, "score"),        // Behavior attached to sprite
    MOVIE(3, "movie"),        // Movie script
    PARENT(7, "parent"),      // Parent script (object factory)
    UNKNOWN(255, "unknown");

    private final int code;
    private final String name;

    ScriptType(int code, String name) {
        this.code = code;
        this.name = name;
    }

    public int getCode() {
        return code;
    }

    public String getName() {
        return name;
    }

    public static ScriptType fromCode(int code) {
        for (ScriptType type : values()) {
            if (type.code == code) return type;
        }
        return UNKNOWN;
    }

    public boolean isBehavior() {
        return this == SCORE;
    }

    public boolean isMovieScript() {
        return this == MOVIE;
    }

    public boolean isParentScript() {
        return this == PARENT;
    }

    @Override
    public String toString() {
        return name;
    }
}

File: cpp/src/cast/ScriptType.cpp
#include "libreshockwave/cast/ScriptType.hpp"

namespace libreshockwave::cast {

int code(ScriptType type) {
    return static_cast<int>(type);
}

std::string_view name(ScriptType type) {
    switch (type) {
        case ScriptType::Invalid: return "invalid";
        case ScriptType::Score: return "score";
        case ScriptType::Movie: return "movie";
        case ScriptType::Parent: return "parent";
        case ScriptType::Unknown: return "unknown";
    }
    return "unknown";
}

ScriptType scriptTypeFromCode(int value) {
    switch (value) {
        case 0: return ScriptType::Invalid;
        case 1: return ScriptType::Score;
        case 3: return ScriptType::Movie;
        case 7: return ScriptType::Parent;
        default: return ScriptType::Unknown;
    }
}

bool isBehavior(ScriptType type) { return type == ScriptType::Score; }
bool isMovieScript(ScriptType type) { return type == ScriptType::Movie; }
bool isParentScript(ScriptType type) { return type == ScriptType::Parent; }

} // namespace libreshockwave::cast
